### PR TITLE
Add repeater rate limiter based on number of attempts

### PR DESCRIPTION
**Pull request by gherceg**
## Technical Summary

When a project has both slow repeater endpoints _and_ a lot of records to forward, our workers are unable to keep up with the backlog despite aggressive rate limiting. This is because in order to rate limit based on wait time, we (HQ) has to wait for the repeater attempts to return before reporting the amount of time it took to the rate limiter. This means as soon as the window opens up for a project with high volume to attempt forwarding, most of the workers will pickup a task for that project, and start attempt to hit the slow endpoint simultaneously, leaving no room for another project for potentially up to 5 minutes (repeater attempt request timeout). 

By combining this rate limiting technique with rate limiting the sheer number of attempts, we can report the usage of an attempt before actually sending the request, which should make it so that more requests from the specific problematic project are rate limited.

The downside of this code is that it could impact projects that have fast endpoints but high volume, since as soon as the global rate limit thresholds are exceeded, surpassing either the project level wait times or attempt limits will result in being rate limited.

### Safety story

The functional change to rate limiting is stuck behind a feature flag, giving us the ability to completely bypass this change if needed.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

File: corehq/motech/rate_limiter.py [hunk 1]
```diff
@@ -55,14 +60,34 @@ def _get_per_user_repeater_wait_milliseconds_rate_definition(domain):
...
+    return PerUserRateDefinition(
+        per_user_rate_definition=get_dynamic_rate_definition(
+            "repeater_attempts_per_user",
+            default=RateDefinition(
+                per_week=None,
+                per_day=None,
+                per_hour=6000,
+                per_minute=100,
+                per_second=10,
+            ),
```
Comment (AmitPhulera) [inline corehq/motech/rate_limiter.py:74]: Does this mean that we will throttle after a domain makes more than 10 req/sec or 100 req/min or 6000 req/hour?

Comment (gherceg) [reply corehq/motech/rate_limiter.py:74]: Yes, if the global wait times are exceeded. These can also be tuned after the fact, and for that reason should be set pretty conservatively, so bumping them higher doesn't sound like a bad idea if that is your concern.

Comment (AmitPhulera) [reply corehq/motech/rate_limiter.py:74]: No. Its fine. I was just asking for my understanding that it is the way it is supposed to work.

Comment (gherceg) [reply corehq/motech/rate_limiter.py:74]: Oh but this is a per user definition, so it would actually be much higher. In this case, I wonder if it is better to just use a simple RateDefinition.

Comment (gherceg) [reply corehq/motech/rate_limiter.py:74]: I think for the goal of this PR, I'm leaning towards switching this to a RateDefinition, otherwise I'm worried large projects could still avoid being rate limited. I view this as the only way we have of limiting the number of workers a project can use at a time, to ensure that other workers remain free to continue processing other records.

File: corehq/toggles/__init__.py [hunk 2]
```diff
@@ -1982,6 +1982,17 @@ def _commtrackify(domain_name, toggle_is_enabled):
...
 )
 
+RATE_LIMIT_REPEATER_ATTEMPTS = DynamicallyPredictablyRandomToggle(
+    'rate_limit_repeater_attempts',
+    'Apply rate limiting to attempts for data forwarding (repeaters)',
+    TAG_INTERNAL,
+    [NAMESPACE_DOMAIN],
+    description="""
+    In addition to the rate limits based on time spent waiting, these rate limits ensure a project
+    is limited to how many workers they can use at a time.
```
Comment (AmitPhulera) [inline corehq/toggles/__init__.py:1992]: When I read the description, it looks like we are limiting the number of workers  a particular domain can use. I think which is not the case.
Can we rephrase it to something like rate limiting the volume of requests that a domain can make?

File: corehq/motech/rate_limiter.py (hunk 1 above)
Comment (AmitPhulera) [reply corehq/motech/rate_limiter.py]: > I view this as the only way we have of limiting the number of workers a project can use at a time.

I think I might have not understood the changes correctly but where exactly are we limiting the number of workers?

File: corehq/motech/rate_limiter.py [hunk 3]
```diff
@@ -87,6 +112,13 @@ def report_repeater_usage(domain, milliseconds):
     _report_current_global_repeater_thresholds()
 
 
+@run_only_when(SHOULD_RATE_LIMIT_REPEATERS)
```
[the remark this answers is not in the document]
Comment (gherceg) [reply corehq/motech/rate_limiter.py:112]: Good point. It looks like the decorator [can handle either a callable or not](https://github.com/dimagi/commcare-hq/blob/73de2c1c8405d1dd9986585c55215195480147a5/corehq/util/decorators.py#L55-L63), but I can remove the unnecessary lambda from line 82.

File: corehq/toggles/__init__.py (hunk 2 above)
Comment (gherceg) [reply corehq/toggles/__init__.py]: Yeah, conceptually we are limiting the number of workers, but I understand why that isn't helpful. Updated in 744f7a7b8bd7ec5d4fd3d9378476879df075912b

File: corehq/motech/rate_limiter.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gherceg) [reply corehq/motech/rate_limiter.py:113]: babb4d6989abccf112c62abc5ec5146ca1cbc4d5

File: corehq/motech/rate_limiter.py [hunk 4]
```diff
@@ -55,14 +60,34 @@ def _get_per_user_repeater_wait_milliseconds_rate_definition(domain):
 )
 
 
+def _get_per_user_repeater_attempt_rate_definition(domain):
```
[the remark this answers is not in the document]
Comment (gherceg) [reply corehq/motech/rate_limiter.py]: ffdaa24e51d7787593a477e4ee15ed0ac98ac6f0

File: corehq/motech/rate_limiter.py (hunk 1 above)
Comment (gherceg) [reply corehq/motech/rate_limiter.py]: Yeah sorry I mentioned this the comment on the feature flag description as well. We aren't explicitly limiting the number of workers, but conceptually are (I think). The idea is that by limiting the number of records a domain can attempt to forward in a given time window, we are limiting the number of workers they will use in that time window too. If a domain has records that take ~a minute to send to an endpoint, and is able to attempt records for all 100 celery workers (let's pretend 100 is the total concurrency), for that minute, all celery workers will be busy attempting to forward records for that domain. If we instead say a domain can only attempt 10 forwards a minute, that domain is only able to use 10% of our total concurrency/number of workers.

The numbers are obviously way off, but that is the idea.

File: corehq/motech/repeaters/tasks.py [hunk 5]
```diff
@@ -170,6 +174,7 @@ def _process_repeat_record(repeat_record):
             # with the intent of avoiding clumping and spreading load
             repeat_record.postpone_by(random.uniform(*RATE_LIMITER_DELAY_RANGE))
         elif repeat_record.is_queued():
+            report_repeater_attempt(repeat_record.domain)
```
Comment (gherceg) [inline corehq/motech/repeaters/tasks.py:177]: To me, this is the crux of this change. Having something that reports _before_ we attempt to forward is very valuable when dealing with slow endpoints, since by the time the slow endpoints return and we report wait times, our rate limiting code is taking action too late. By reporting before, we can prevent too many workers from actually attempting to forward to a slow endpoint.

**Review by kaapstorm — APPROVED** [review]
I'd support renaming `RATE_LIMIT_REPEATERS` to something like `RATE_LIMIT_REPEATER_WAIT` to disambiguate these.

**Review by millerdev — APPROVED** [review]

**Comment by gherceg** [comment]
Immediate regrets. Reverting cause there was an issue with `_get_repeater_attempt_rate_definition` because the `get_rate_limits` method in a RateLimiter expects to have one parameter to represent the scope. I'll add that change to a new PR and apologies for the mess.
